extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/MyPermission.java

```diff
@@ -0,0 +1,47 @@
+package io.quarkus.resteasy.reactive.server.test.security;
+
+import java.security.Permission;
+import java.util.Objects;
+
+public class MyPermission extends Permission {
+
+    static final MyPermission EMPTY = new MyPermission("my-perm", null, null);
+
+    private final String authorization;
+    private final String queryParam;
+
+    public MyPermission(String permissionName, String authorization, String queryParam) {
+        super(permissionName);
+        this.authorization = authorization;
+        this.queryParam = queryParam;
+    }
+
+    @Override
+    public boolean implies(Permission permission) {
+        if (permission instanceof MyPermission myPermission) {
+            return myPermission.authorization != null && "query1".equals(myPermission.queryParam);
```

Comment:
Once again a permission where `this.` members are never used. I don't understand this model.

Reply:
The model comes from the https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/security/Permission.html#implies(java.security.Permission). In Quarkus terms, the identity has permission checker that can check required permissions. Here it is:

- https://github.com/quarkusio/quarkus-security/blob/8e9154b14121f692cbad186a9ec7722d3baaff64/src/main/java/io/quarkus/security/identity/SecurityIdentity.java#L121

So technically, you only need one _required_ permission and one checker. You can do `requiredPermission.implies(null)` or whatever, but it feels very hacky, so I always expect users will declare both `requiredPermission` and `possessedPermission` and do: `possessedPermission.implies(requiredPermission)`.

I expect this can be simplified with your `@RequiredPermission` or `@PermissionChecker` or whatever later.

BTW this particular `MyPermission` is here to test that bean parameters were passed correctly, I don't think it needs to use `this` for that purpose. I wouldn't write this to the docs.

Reply:
I can understand `implies` as it's documented, which is about figuring out if, given an already granted permission, another one is implied by it. For example, given that the current user has the `admin` permission, is the `read` permission implied? Probably, yes.

But this is not what we're describing here with `MyPermission`. I see no relation.

`SecurityIdentity.checkPermission(Permission)`  I also understand: does the current user have the right to do X? Fine. That makes sense.

But again, this does not look like what `MyPermission.implies` is doing. Is this something related to the current PR, or the general security model in Quarkus?

Reply:
> But this is not what we're describing here with MyPermission. I see no relation.

ad `describing`: it's a test, not docs; but sure, I'll answer

> SecurityIdentity.checkPermission(Permission) I also understand: does the current user have the right to do X? Fine. That makes sense.

When you specify `@PermissionsAllowed(permission=CustomPermission)` then `CustomPermission` instance is created and passed to the `SecurityIdentity.checkPermission(Permission)` because it is fact `checkPermission(requiredPermission)`.

It depends on how user writes `SecurityIdentityAugmentor` whether `this` is used. I am using one and the same permission in the all tests for making it easy, but you can also do this (meta code ...):

```
@Override
public Uni<SecurityIdentity> augment(SecurityIdentity identity, AuthenticationRequestContext context) {
    QuarkusSecurityIdentity.builder(identity).addPermissionChecker(requiredPermission -> {
        if (requiredPermission.implies(new StringPermission("read"))) {
            return Uni.createFrom().item(Boolean.TRUE);
        }
        return Uni.createFrom().item(Boolean.FALSE);
    }).build();
}
```

and this augmentor part can be simplified with https://github.com/quarkusio/quarkus/issues/43238.

I cannot simplify this model in this PR.

Reply:
Oh, so wait, users have to implement a `SecurityIdentityAugmentor` to make this work?

Reply:
It is one option. There are other ways to add `permission checker` on your security identity. For example we map token claims to StringPermission. Or you can remap SecurityIdentity roles like this:

# map identity role 'user' to String permission 'read'
quarkus.http.auth.policy.role-policy2.permissions.user=read  
quarkus.http.auth.permission.roles2.paths=*
quarkus.http.auth.permission.roles2.policy=role-policy2

Which gives you `var possessed = new StringPermission("read")`

Problem with this is that we add permission check like `possessedPermisison.implies(requiredPermission)`. It felt right, but I suppose for purpose of this particular situation, `possessedPermisison.implies(requiredPermission)` is equivalent. Though it is not from POV of formal logic.

So ATM you do need augmentor, but this part could be changed in this PR if it is requested.

Reply:
It would be yet another breaking change though.

Reply:
And if you wonder, if user doesn't specify any permission checker but `@PermissionsAllowed` is applied, the `QuarkusSecurityIdentity` responds with false by default https://github.com/quarkusio/quarkus/blob/1603d2cd825289f1fbf744447ca3bf87237df889/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java#L84

so you don't have the permission unless you posses at least one that complies it. I think we shouldn't change it in this PR. I'll continue updating the docs with the augmentor.